File: quickstart_guides/linked_lists/java/IsPalindrome.java
/*
 * Title: Is palindrome
 *
 * Problem:
 *  Check if linked list is a palindrome.
 *
 * Execution: javac IsPalindrome.java && java IsPalindrome
 */
import java.util.*;


class IsPalindrome {
    static ListNode head;

    public static class ListNode {
        int val;
        ListNode next;
        ListNode() {}
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }

    public static boolean isPalindrome(ListNode head) {
        if(head!=null) {
            List<Integer> s=new ArrayList<>();
            ListNode c=head;

            while(c!=null){
                s.add(c.val);
                c=c.next;
            }

            boolean isPalindrome=true;
            for(int i=0;i<s.size()/2+1;i++){
                if(s.get(i)!=s.get(s.size()-i-1)){
                    isPalindrome=false;
                    break;
                }
            }

            return isPalindrome;
        }else{
            return true;
        }
    }


    public static void main(String[] args) {
        IsPalindrome list1 = new IsPalindrome();
        list1.head = new ListNode(1);
        list1.head.next = new ListNode(2);

        assert list1.isPalindrome(list1.head) == false;

        IsPalindrome list2 = new IsPalindrome();
        list2.head = new ListNode(1);
        list2.head.next = new ListNode(2);
        list2.head.next.next = new ListNode(2);
        list2.head.next.next.next = new ListNode(1);

        assert list2.isPalindrome(list2.head) == true;

        System.out.println("All tests have passed.");

    }
}

File: quickstart_guides/linked_lists/python/is_palindrome.py
"""
Title: Linked list is palindrome

Problem:
    Check if linked list has nodes that consist of palindrome.

Execution: python is_palindrome.py
"""
import unittest


class ListNode:
    """Class for Linked List node."""

    def __init__(self, x: int) -> None:
        self.val = x
        self.next = None


def is_palindrome(head: ListNode) -> bool:
    """Check if linked list is palindrome."""
    vals = []
    curr = head
    while curr:
        vals.append(curr.val)
        curr = curr.next

    while head:
        if head.val != vals.pop():
            return False
        head = head.next
    return True


class TestIsPalindrome(unittest.TestCase):
    """Unit test for is_palindrome."""

    def test_1(self):
        list_1 = ListNode(1)
        list_1.next = ListNode(2)

        self.assertEqual(is_palindrome(list_1), False)

    def test_2(self):
        list_2 = ListNode(1)
        list_2.next = ListNode(2)
        list_2.next.next = ListNode(2)
        list_2.next.next.next = ListNode(1)

        self.assertEqual(is_palindrome(list_2), True)


if __name__ == "__main__":
    unittest.main()
